// panda-parchment/build/npm/src/time.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.times = exports.benchmark = exports.microseconds = exports.timer = exports.sleep = void 0;

var _pandaGarden = require("panda-garden");

var _pandaGenerics = _interopRequireDefault(require("panda-generics"));

var _promise = require("./promise");

var _type = require("./type");

var _string = require("./string");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var benchmark, create, define, microseconds, sleep, timer, times;
exports.times = times;
exports.timer = timer;
exports.sleep = sleep;
exports.microseconds = microseconds;
exports.benchmark = benchmark;
({
  create,
  define
} = _pandaGenerics.default);

exports.timer = timer = function (wait, action) {
  var id;
  id = setTimeout(action, wait);
  return function () {
    return clearTimeout(id);
  };
};

exports.sleep = sleep = function (interval) {
  return (0, _promise.promise)(function (resolve, reject) {
    return timer(interval, function () {
      return resolve();
    });
  });
};

exports.times = times = (0, _pandaGarden.curry)(function (fn, n) {
  var results;
  results = [];

  while (n-- !== 0) {
    results.push(fn());
  }

  return results;
});

exports.microseconds = microseconds = function () {
  var ref;

  if ((typeof process !== "undefined" && process !== null ? process.hrtime : void 0) != null) {
    return function () {
      var nanoseconds, seconds;
      [seconds, nanoseconds] = process.hrtime();
      return seconds * 1000000 + nanoseconds / 1000;
    };
  } else if ((typeof window !== "undefined" && window !== null ? (ref = window.performance) != null ? ref.now : void 0 : void 0) != null) {
    return function () {
      return window.performance.now();
    };
  } else {
    return function () {
      return Date.now() * 1000;
    };
  }
}();

exports.benchmark = benchmark = create({
  name: "benchmark",
  description: "Times a function's execution with up to microsecond resolution"
});
define(benchmark, _type.isFunction, function (fn) {
  var start;
  start = microseconds();
  fn();
  return microseconds() - start;
});
define(benchmark, _type.isAsyncFunction, async function (fn) {
  var start;
  start = microseconds();
  await fn();
  return microseconds() - start;
});
