// panda-generics/build/npm/src/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var GenericFunction,
    create,
    define,
    lookup,
    splice = [].splice;
GenericFunction = class GenericFunction {
  static create(options = {}) {
    return new this(options);
  }

  constructor({
    name,
    description,
    default: _default
  }) {
    this.name = name;
    this.description = description;
    this.default = _default;

    if (this.name == null) {
      this.name = "anonymous-generic";
    }

    this.entries = [];

    if (this.default == null) {
      this.default = (...args) => {
        var error;
        error = new TypeError(`${this.name}: Invalid arguments.`);
        error.arguments = args;
        throw error;
      };
    }
  }

  define(...terms) {
    var f, ref;
    ref = terms, [...terms] = ref, [f] = splice.call(terms, -1);
    return this.entries.unshift([terms, f]);
  }

  lookup(args) {
    var f, i, j, len, match, ref, term, terms;
    ref = this.entries; // go through each definition in our lookup 'table'

    for (j = 0, len = ref.length; j < len; j++) {
      [terms, f] = ref[j];

      if (terms.length > args.length) {
        // there must be at least one argument per term
        // (variadic terms can consume multiple arguments,
        // so the converse is not true)
        continue;
      } // we can't have a match if we don't match any terms


      match = false; // each argument must be consumed

      i = 0;

      while (i < args.length) {
        if ((term = terms[i]) == null) {
          match = false;
          break;
        } // if the term may be variadic (indicated by taking 0 arguments)
        // try the term with the remaining arguments


        if (term.length === 0) {
          match = term(...args.slice(i));
          break;
        }

        if (!(match = term(args[i++]))) {
          // otherwise, we have the default case, where we try to match
          // the next argument with the next term
          break;
        }
      }

      if (match) {
        // if we ended up with a match, just return the corresponding fn
        return f;
      }
    } // if exit the loop without returning a match, return the default


    return this.default;
  }

  dispatch(args) {
    var f;
    f = this.lookup(args);
    return f(...args);
  }

};

lookup = function ({
  _
}, args) {
  return _.lookup(args);
};

define = function ({
  _
}, ...args) {
  return _.define(...args);
};

create = function (options) {
  var generic;

  generic = function (...args) {
    return generic._.dispatch(args);
  };

  generic._ = GenericFunction.create(options);
  return generic;
};

var _default2 = {
  create,
  define,
  lookup
};
exports.default = _default2;
