// panda-river/build/npm/src/iterator.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isDone = exports.value = exports.next = exports.isIterator = exports.iterator = exports.isIterable = void 0;

var _pandaGarden = require("panda-garden");

var _pandaGenerics = _interopRequireDefault(require("panda-generics"));

var _pandaParchment = require("panda-parchment");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var create, define, isDone, isIterable, isIterator, iterator, next, value;
exports.value = value;
exports.next = next;
exports.iterator = iterator;
exports.isIterator = isIterator;
exports.isIterable = isIterable;
exports.isDone = isDone;
({
  create,
  define
} = _pandaGenerics.default);

exports.isIterable = isIterable = function (x) {
  return (0, _pandaParchment.isFunction)(x != null ? x[Symbol.iterator] : void 0) || (0, _pandaParchment.isGeneratorFunction)(x);
};

exports.isIterator = isIterator = function (x) {
  return (0, _pandaParchment.isFunction)(x != null ? x.next : void 0) && isIterable(x);
};

exports.iterator = iterator = create({
  name: "iterator",
  description: "produces an iterator from an input",
  default: "unable to create iterator from value"
});
define(iterator, _pandaParchment.isFunction, function (f) {
  return {
    next: f,
    [Symbol.iterator]: function () {
      return this;
    }
  };
});
define(iterator, isIterable, function (i) {
  return i[Symbol.iterator]();
});
define(iterator, _pandaParchment.isGeneratorFunction, function (g) {
  return g();
});
define(iterator, isIterator, function (i) {
  return i;
});

exports.next = next = function (i) {
  return i.next();
};

exports.value = value = function ({
  value
}) {
  return value;
};

exports.isDone = isDone = function ({
  done
}) {
  return done;
};
