// panda-river/build/web/src/iterator.js
var create, define, isDone, isIterable, isIterator, iterator, next, value;
import { identity, curry } from "panda-garden";
import Method from "panda-generics";
import { isKind, isFunction, isGeneratorFunction } from "panda-parchment";
({
  create,
  define
} = Method);

isIterable = function (x) {
  return isFunction(x != null ? x[Symbol.iterator] : void 0) || isGeneratorFunction(x);
};

isIterator = function (x) {
  return isFunction(x != null ? x.next : void 0) && isIterable(x);
};

iterator = create({
  name: "iterator",
  description: "produces an iterator from an input",
  default: "unable to create iterator from value"
});
define(iterator, isFunction, function (f) {
  return {
    next: f,
    [Symbol.iterator]: function () {
      return this;
    }
  };
});
define(iterator, isIterable, function (i) {
  return i[Symbol.iterator]();
});
define(iterator, isGeneratorFunction, function (g) {
  return g();
});
define(iterator, isIterator, function (i) {
  return i;
});

next = function (i) {
  return i.next();
};

value = function ({
  value
}) {
  return value;
};

isDone = function ({
  done
}) {
  return done;
};

export { isIterable, iterator, isIterator, next, value, isDone };
