// panda-parchment/build/npm/src/object.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.fromJSON = exports.toJSON = exports.merge = exports.extend = exports.include = exports.assign = exports.omit = exports.pick = exports.pairs = exports.values = exports.keys = exports.has = exports.methods = exports.properties = exports.detach = exports.bind = exports.property = void 0;

var _pandaGarden = require("panda-garden");

var _type = require("./type");

var assign, bind, detach, extend, fromJSON, has, include, keys, merge, methods, omit, pairs, pick, properties, property, toJSON, values;
exports.values = values;
exports.toJSON = toJSON;
exports.property = property;
exports.properties = properties;
exports.pick = pick;
exports.pairs = pairs;
exports.omit = omit;
exports.methods = methods;
exports.merge = merge;
exports.keys = keys;
exports.include = include;
exports.has = has;
exports.fromJSON = fromJSON;
exports.extend = extend;
exports.detach = detach;
exports.bind = bind;
exports.assign = assign;
exports.property = property = (0, _pandaGarden.curry)(function (key, object) {
  return object[key];
});
exports.bind = bind = (0, _pandaGarden.curry)(function (f, x) {
  return f.bind(x);
});

exports.detach = detach = function (f) {
  return (0, _pandaGarden.curry)(function (x, ...args) {
    return f.apply(x, args);
  });
};

exports.properties = properties = function (target, descriptions) {
  var description, name, results;
  results = [];

  for (name in descriptions) {
    description = descriptions[name];

    if (description.enumerable == null) {
      description.enumerable = true;
    }

    if (description.configurable == null) {
      description.configurable = true;
    }

    results.push(Object.defineProperty(target, name, description));
  }

  return results;
};

exports.methods = methods = function (target, definitions) {
  var f, name, results;
  results = [];

  for (name in definitions) {
    f = definitions[name];
    results.push(Object.defineProperty(target, name, {
      value: f,
      configurable: true,
      writable: true
    }));
  }

  return results;
};

exports.has = has = (0, _pandaGarden.curry)(function (p, x) {
  return x[p] != null;
});
exports.keys = keys = Object.keys;

exports.values = values = function (x) {
  var k, results, v;
  results = [];

  for (k in x) {
    v = x[k];
    results.push(v);
  }

  return results;
};

exports.pairs = pairs = function (x) {
  var k, results, v;
  results = [];

  for (k in x) {
    v = x[k];
    results.push([k, v]);
  }

  return results;
};

exports.pick = pick = (0, _pandaGarden.curry)(function (f, x) {
  var k, r, v;
  r = {};

  for (k in x) {
    v = x[k];

    if (f(k, v)) {
      r[k] = v;
    }
  }

  return r;
});
exports.omit = omit = (0, _pandaGarden.curry)(function (f, x) {
  return pick((0, _pandaGarden.negate)(f), x);
});

exports.include = include = exports.extend = extend = exports.assign = assign = function (target, ...sources) {
  return Object.assign(target, ...sources);
};

exports.merge = merge = function (...objects) {
  return Object.assign({}, ...objects);
};

exports.toJSON = toJSON = function (x, pretty = false) {
  if (pretty) {
    return JSON.stringify(x, null, 2);
  } else {
    return JSON.stringify(x);
  }
};

exports.fromJSON = fromJSON = JSON.parse;
